File: Multi-Way Trees/b-star-tree-implementation.py
# File: b-star-tree-implementation.py
# Program implementing B* Tree operations

class Node:

    def __init__(self):
        self.keys = []
        self.children = []
        self.leaf = True


class BStarTree:

    def __init__(self):
        self.root = Node()

    def traverse(self,node):

        for i in range(len(node.keys)):

            if not node.leaf:
                self.traverse(node.children[i])

            print(node.keys[i], end=" ")

        if not node.leaf:
            self.traverse(node.children[len(node.keys)])

    def search(self,node,key):

        i=0

        while i < len(node.keys) and key > node.keys[i]:
            i+=1

        if i < len(node.keys) and key == node.keys[i]:
            return node

        if node.leaf:
            return None

        return self.search(node.children[i],key)

    def insert(self,key):

        if len(self.root.keys) < 4:
            self.root.keys.append(key)
            self.root.keys.sort()


if __name__ == "__main__":

    tree = BStarTree()

    tree.insert(10)
    tree.insert(20)
    tree.insert(5)
    tree.insert(30)
    tree.insert(15)

    print("Traversal of B* Tree:")

    tree.traverse(tree.root)

File: Multi-Way Trees/b-star-tree-implementation.java
// File: b-star-tree-implementation.java
// Program implementing B* Tree operations

class Node {

    int keys[] = new int[4];
    Node children[] = new Node[5];
    int numKeys = 0;
    boolean leaf = true;
}

public class BStarTree {

    Node root = new Node();

    void traverse(Node node)
    {
        if(node==null)
            return;

        int i;

        for(i=0;i<node.numKeys;i++)
        {
            if(!node.leaf)
                traverse(node.children[i]);

            System.out.print(node.keys[i]+" ");
        }

        if(!node.leaf)
            traverse(node.children[i]);
    }

    Node search(Node node,int key)
    {
        int i=0;

        while(i<node.numKeys && key>node.keys[i])
            i++;

        if(i<node.numKeys && key==node.keys[i])
            return node;

        if(node.leaf)
            return null;

        return search(node.children[i],key);
    }

    void insert(int key)
    {
        if(root.numKeys < 4)
        {
            root.keys[root.numKeys++] = key;
        }
    }

    public static void main(String args[])
    {
        BStarTree tree = new BStarTree();

        tree.insert(10);
        tree.insert(20);
        tree.insert(5);
        tree.insert(30);

        System.out.println("Traversal of B* Tree:");

        tree.traverse(tree.root);
    }
}
